### PR TITLE
BufferGeometryUtils: Add "deinterleaveAttribute", "deinterleaveGeometry" functions

### PR DESCRIPTION
**Pull request by gkjohnson**
Related issue: --

**Description**

Adds two functions to produce deinterleaved attributes and deinterleave all attributes on a particular geometry in place. This is useful because there are some scenarios where working with interleaved attributes can be a bit of a pain. Specifically "mergeBufferGeometries" does not support interleaved attributes (though maybe that could be updated) and it's risky to transfer an interleaved attribute buffer to a webworker since it may wind up neutering other interleaved attribute buffer arrays which is otherwise not super traceable.

Both uses cases are what I'm concerned with for the current path tracing work since buffer geometries must be merged and the index and position attribute arrays are transferred to a web worker to be used for generating a BVH.

### CODE REVIEW

File: examples/jsm/utils/BufferGeometryUtils.js [hunk 1]
```diff
@@ -415,6 +415,75 @@ function interleaveAttributes( attributes ) {
...
+
+		}
+
+	}
+
+	if ( geometry.index && geometry.index.isInterleavedBufferAttribute ) {
+
+		geometry.index = deinterleaveAttribute( geometry.index );
+
+	}
```
Comment (donmccurdy) [inline examples/jsm/utils/BufferGeometryUtils.js:483]: I don't think interleaved indices are a thing ([source](https://stackoverflow.com/questions/11148567/rendering-meshes-with-multiple-indices/11148568#11148568)), so we could skip a copy here.

File: examples/jsm/utils/BufferGeometryUtils.js [hunk 2]
```diff
@@ -415,6 +415,75 @@ function interleaveAttributes( attributes ) {
...
+
+	for ( const key in geometry.attributes ) {
+
+		if ( geometry.attributes[ key ].isInterleavedBufferAttribute ) {
+
+			geometry.attributes[ key ] = deinterleaveAttribute( geometry.attributes[ key ] );
+
+		}
+
+	}
```
Comment (donmccurdy) [inline examples/jsm/utils/BufferGeometryUtils.js:477]: Would you mind throwing morph targets in too? 🙏

File: examples/jsm/utils/BufferGeometryUtils.js (hunk 1 above)
Comment (gkjohnson) [reply examples/jsm/utils/BufferGeometryUtils.js:489]: gotcha -- done!

File: examples/jsm/utils/BufferGeometryUtils.js (hunk 2 above)
Comment (gkjohnson) [reply examples/jsm/utils/BufferGeometryUtils.js:479]: Added morph targets in, as well. Is there ever a case where the attributes object and the morphTargets object could reference the same BufferAttribute object instance?

Comment (donmccurdy) [reply examples/jsm/utils/BufferGeometryUtils.js:479]: Not in glTF, where the morphs are relative deltas. It wouldn't seem crazy to reuse the attribute for absolute targets (like morph normals or tangents). The more common case would be `uv` and `uv2` sharing the same attribute on the base geometry.

Comment (gkjohnson) [reply examples/jsm/utils/BufferGeometryUtils.js:479]: > The more common case would be uv and uv2 sharing the same attribute on the base geometry.

Oh good point. I can add a map so any existing reused attributes continued to be structured that way. How would that look? Would they be literally the same BufferAttribute instance (ie `attributes.uv === attributes.uv2`)? Or just the same array / InterleavedBuffer instance?

Comment (donmccurdy) [reply examples/jsm/utils/BufferGeometryUtils.js:479]: For glTF they'd be the same BufferAttribute instance – I think that's a good enough proxy here too.

Comment (gkjohnson) [reply examples/jsm/utils/BufferGeometryUtils.js:479]: Great! Done.

**Review by Mugen87 — APPROVED** [review]

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**